### PR TITLE
Add sync cdp interest method

### PR DESCRIPTION
**Pull request by karzak**
Ready for initial review

* adds `SynchronizeInterest` method that gets called whenever a user interacts with a CDP
* adds `CalculateNewInterest` method that can be called by clients to calculate how much interest has accrued since `SynchronizeInterest` was last called on a CDP.

Open questions:

* Because CDP fees are calculated on all principal in the begin blocker, there will be rounding errors between the sum of all CDP debt and the debt in the system. Are there any methods that could break because of this?

### CODE REVIEW

File: x/cdp/keeper/interest.go [hunk 1]
```diff
@@ -86,129 +110,41 @@ func CalculateInterestFactor(perSecondInterestRate sdk.Dec, secondsElapsed sdk.I
...
+	k.SetCdpAndCollateralRatioIndex(ctx, cdp, collateralToDebtRatio)
+	return cdp
 }
 
-// SetTotalPrincipal sets the total amount of principal that has been drawn for the input collateral
-func (k Keeper) SetTotalPrincipal(ctx sdk.Context, collateralType, principalDenom string, total sdk.Int) {
-	store := prefix.NewStore(ctx.KVStore(k.key), types.PrincipalKeyPrefix)
-	_, found := k.GetCollateralTypePrefix(ctx, collateralType)
+// CalculateNewInterest returns the amount of interest that has accrued to the cdp since its interest was last synchronized
+func (k Keeper) CalculateNewInterest(ctx sdk.Context, cdp types.CDP) sdk.Coin {
```
Comment (denalimarsh) [inline x/cdp/keeper/interest.go:139]: Looks like this method is only used in `interest_test.go` and `operations.go` when we need to query a CDP's outstanding interest as CDP queries directly to the store now return out-of-date information. The front-end is going to need to display up-to-date CDP info as well; we should call this function at the start of `LoadAugmentedCDP()` and add the new fees accumulated.

**Review by denalimarsh — COMMENTED** [review]
LGTM. Functions `DepositCollateral`/`WithdrawCollateral` could use some comments to make the justification behind the current ordering a bit clearer.

**Review by rhuairahrighairidh — CHANGES_REQUESTED** [review]
Looks good, left some questions. I'm still looking into the rounding a bit more.

`InitGenesis` and `LiquidateCdps` look they need updating for the new interest format. Are these for a follow up PR?

nit: In `keep/interest.go` in `AccumulateInterest` line 25, the time elapsed is calculated by subtracting unix timestamps. Looks like `.Unix()` truncates times rather than rounding so the elapsed time could be off by up to a second. https://play.golang.org/p/C0saCKvG3A_0 Could subtract the times first then round to second amount.

**Comment by rhuairahrighairidh** [comment]
FYI looks like `sdk.RPow` is slightly inaccurate.
In `interest_test.go`, `TestCalculateInterestFactor`, test "1 year: high interest rate": the calculated rate is different from what wolfram alpha gives https://www.wolframalpha.com/input/?i=%281000000055555%2F1000000000000%29%5E31536000

```
test result: 5.766022095987868825
  wa result: 5.766022096035290485
       diff: 0.000000000047422000
```

**Comment by karzak** [comment]
> `InitGenesis` and `LiquidateCdps` look they need updating for the new interest format. Are these for a follow up PR?

Yes. `InitGenesis` is cleaned up in #738 and I'll address liquidations in a follow up.

File: x/cdp/keeper/cdp.go [hunk 2]
```diff
@@ -41,7 +42,12 @@ func (k Keeper) AddCdp(ctx sdk.Context, owner sdk.AccAddress, collateral sdk.Coi
 
 	// send coins from the owners account to the cdp module
 	id := k.GetNextCdpID(ctx)
-	cdp := types.NewCDP(id, owner, collateral, collateralType, principal, ctx.BlockHeader().Time, sdk.OneDec())
+	interestFactor, found := k.GetInterestFactor(ctx, collateralType)
+	if !found {
+		k.SetInterestFactor(ctx, collateralType, sdk.OneDec())
+		interestFactor = sdk.OneDec()
```
[the remark this answers is not in the document]
Comment (karzak) [reply x/cdp/keeper/cdp.go]: 19e93cc

File: x/cdp/keeper/interest.go [hunk 3]
```diff
@@ -86,129 +110,41 @@ func CalculateInterestFactor(perSecondInterestRate sdk.Dec, secondsElapsed sdk.I
...
-	return newFees
-}
-
-// IncrementTotalPrincipal increments the total amount of debt that has been drawn with that collateral type
-func (k Keeper) IncrementTotalPrincipal(ctx sdk.Context, collateralType string, principal sdk.Coin) {
-	total := k.GetTotalPrincipal(ctx, collateralType, principal.Denom)
-	total = total.Add(principal.Amount)
-	k.SetTotalPrincipal(ctx, collateralType, principal.Denom, total)
-}
+		k.SetCDP(ctx, cdp)
```
[the remark this answers is not in the document]
Comment (karzak) [reply x/cdp/keeper/interest.go:121]: 19e93cc

File: x/cdp/keeper/interest.go (hunk 1 above)
Comment (karzak) [reply x/cdp/keeper/interest.go:136]: 19e93cc

File: x/cdp/keeper/interest.go [hunk 4]
```diff
@@ -86,129 +110,41 @@ func CalculateInterestFactor(perSecondInterestRate sdk.Dec, secondsElapsed sdk.I
...
-	store := prefix.NewStore(ctx.KVStore(k.key), types.PrincipalKeyPrefix)
-	_, found := k.GetCollateralTypePrefix(ctx, collateralType)
+// CalculateNewInterest returns the amount of interest that has accrued to the cdp since its interest was last synchronized
+func (k Keeper) CalculateNewInterest(ctx sdk.Context, cdp types.CDP) sdk.Coin {
+	globalInterestFactor, found := k.GetInterestFactor(ctx, cdp.Type)
 	if !found {
-		panic(fmt.Sprintf("collateral not found: %s", collateralType))
+		return sdk.NewCoin(cdp.AccumulatedFees.Denom, sdk.ZeroInt())
+	}
+	cdpInterestFactor := sdk.OneDec().Add((globalInterestFactor.Sub(cdp.InterestFactor)))
```
[the remark this answers is not in the document]
Comment (karzak) [reply x/cdp/keeper/interest.go]: 19e93cc

File: x/cdp/keeper/interest.go [hunk 5]
```diff
@@ -86,129 +110,41 @@ func CalculateInterestFactor(perSecondInterestRate sdk.Dec, secondsElapsed sdk.I
...
 
-// DecrementTotalPrincipal decrements the total amount of debt that has been drawn for a particular collateral type
-func (k Keeper) DecrementTotalPrincipal(ctx sdk.Context, collateralType string, principal sdk.Coin) {
-	total := k.GetTotalPrincipal(ctx, collateralType, principal.Denom)
-	// NOTE: negative total principal can happen in tests due to rounding errors
-	// in fee calculation
-	total = sdk.MaxInt(total.Sub(principal.Amount), sdk.ZeroInt())
-	k.SetTotalPrincipal(ctx, collateralType, principal.Denom, total)
-}
+	accumulatedInterest := cdp.GetTotalPrincipal().Amount.ToDec().Mul(cdpInterestFactor).RoundInt().Sub(cdp.GetTotalPrincipal().Amount)
```
[the remark this answers is not in the document]
Comment (karzak) [reply x/cdp/keeper/interest.go]: 19e93cc

File: x/cdp/keeper/interest_test.go [hunk 6]
```diff
@@ -318,7 +288,373 @@ func (suite *InterestTestSuite) TestAccumulateInterest() {
...
+			args{
+				ctype:                         "bnb-a",
+				initialTime:                   time.Date(2020, 12, 15, 14, 0, 0, 0, time.UTC),
+				blockInterval:                 7,
+				numberOfBlocks:                10000,
+				initialCDPCollateral:          c("bnb", 10000000000),
+				initialCDPPrincipal:           c("usdx", 500000000),
+				numberOfCdps:                  100,
+				expectedFeesPerCDP:            c("usdx", 54152),
+				exepectedTotalPrincipalPerCDP: c("usdx", 500054152),
```
[the remark this answers is not in the document]
Comment (karzak) [reply x/cdp/keeper/interest_test.go]: 19e93cc

File: x/cdp/keeper/interest_test.go [hunk 7]
```diff
@@ -318,7 +288,373 @@ func (suite *InterestTestSuite) TestAccumulateInterest() {
...
+			suite.Require().Equal(tc.args.expectedTotalPrincipal, totalPrincipal)
+
+			sumOfCDPPrincipal := sdk.ZeroInt()
+
+			for j := 0; j < tc.args.numberOfCdps; j++ {
+				cdp, found := suite.keeper.GetCDP(suite.ctx, tc.args.ctype, uint64(j+1))
+				suite.Require().True(found)
+				cdp = suite.keeper.SynchronizeInterest(suite.ctx, cdp)
+				suite.Require().Equal(tc.args.expectedFeesPerCDP, cdp.AccumulatedFees)
+				suite.Require().Equal(tc.args.exepectedTotalPrincipalPerCDP, cdp.GetTotalPrincipal())
```
[the remark this answers is not in the document]
Comment (karzak) [reply x/cdp/keeper/interest_test.go]: 19e93cc

File: x/cdp/keeper/interest.go [hunk 8]
```diff
@@ -23,7 +22,7 @@ func (k Keeper) AccumulateInterest(ctx sdk.Context, ctype string) error {
 		return nil
 	}
 
-	timeElapsed := ctx.BlockTime().Unix() - previousAccrualTime.Unix()
+	timeElapsed := int64(ctx.BlockTime().Sub(previousAccrualTime).Seconds())
```
Comment (rhuairahrighairidh) [inline x/cdp/keeper/interest.go:25]: nit:
```suggestion
	timeElapsed := int64(math.RoundToEven(
		ctx.BlockTime().Sub(previousAccrualTime).Seconds(),
	))
```

**Review by rhuairahrighairidh — APPROVED** [review]
LGTM, good catch with the index update, I missed that one

**Pull request merged**